packages/frontend/workspace/src/providers/sync/peer.ts

```diff
@@ -0,0 +1,379 @@
+        throw err;
+      }
+    }
+  }
+
+  private state: {
+    connectedDocs: Map<string, Doc>;
+    pushUpdatesQueue: AsyncQueue<{
+      docId: string;
+      data: Uint8Array;
+    }>;
+    pullUpdatesQueue: AsyncQueue<{
+      docId: string;
+      data: Uint8Array;
+    }>;
+    subdocsLoadQueue: AsyncQueue<Doc>;
+  } = {
+    connectedDocs: new Map(),
+    pushUpdatesQueue: new AsyncQueue(),
+    pullUpdatesQueue: new AsyncQueue(),
+    subdocsLoadQueue: new AsyncQueue(),
+  };
+
+  /**
+   * main synchronization logic
+   */
+  async sync(abortOuter: AbortSignal) {
+    const abortInner = new AbortController();
+
+    abortOuter.addEventListener('abort', reason => {
+      abortInner.abort(reason);
+    });
+
+    let dispose: (() => void) | null = null;
+    try {
+      // start listen storage updates
+      dispose = await this.storage.subscribe(
+        this.handleStorageUpdates,
+        reason => {
+          // abort if storage disconnect, should trigger retry loop
+          abortInner.abort('subscribe disconnect:' + reason);
+        }
+      );
+      throwIfAborted(abortInner.signal);
+
+      // Step 1: load root doc
+      this.status = SyncPeerStatus.LoadingRootDoc;
+
+      await this.connectDoc(this.rootDoc, abortInner.signal);
+
+      this.status = SyncPeerStatus.LoadingSubDoc;
+
+      // Step 2: load subdocs
+      this.state.subdocsLoadQueue.push(
+        ...Array.from(this.rootDoc.getSubdocs())
+      );
+
+      this.rootDoc.on('subdocs', this.handleSubdocsUpdate);
+
+      while (this.state.subdocsLoadQueue.length > 0) {
+        const subdoc = await this.state.subdocsLoadQueue.next(
+          abortInner.signal
+        );
+        await this.connectDoc(subdoc, abortInner.signal);
+      }
+
+      this.status = SyncPeerStatus.Syncing;
+      this.updateSyncStatus();
+
+      // Finally: start sync
+      await Promise.all([
+        // listen subdocs
+        (async () => {
+          while (throwIfAborted(abortInner.signal)) {
+            const subdoc = await this.state.subdocsLoadQueue.next(
+              abortInner.signal
+            );
+            this.status = SyncPeerStatus.LoadingSubDoc;
+            await this.connectDoc(subdoc, abortInner.signal);
+            this.status = SyncPeerStatus.Syncing;
+            this.updateSyncStatus();
+          }
+        })(),
+        // pull updates
+        (async () => {
+          while (throwIfAborted(abortInner.signal)) {
+            const { docId, data } = await this.state.pullUpdatesQueue.next(
+              abortInner.signal
+            );
+            this.updateSyncStatus();
+            const subdoc = this.state.connectedDocs.get(docId);
+            if (subdoc) {
+              applyUpdate(subdoc, data, this.name);
+            }
+          }
+        })(),
+        // push updates
+        (async () => {
+          while (throwIfAborted(abortInner.signal)) {
+            const { docId, data } = await this.state.pushUpdatesQueue.next(
```

Comment:
we gonna need to filter out the update with empty uint8Array or `uint8Array([0, 0])`, which is empty doc serialization result

Reply:
<img width="835" alt="image" src="https://github.com/toeverything/AFFiNE/assets/8281226/0998a908-1eac-4df0-862d-26cc849a342d">
batch of useless updates sended